twoliter/embedded/imghelper

```diff
@@ -0,0 +1,398 @@
+
+sbsetup() {
+  local pedb sb_key_source
+  local -n shim_sign_key code_sign_key
+  shim_sign_key="${1:?}"
+  code_sign_key="${2:?}"
+
+  # For an AWS profile, we expect a config file for the PKCS11
+  # helper. Otherwise, there should be a local key and cert.
+  if [[ -s "${HOME}/.config/aws-kms-pkcs11/config.json" ]]; then
+    # Set AWS environment variables from build secrets, if present.
+    for var in AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY AWS_SESSION_TOKEN; do
+      val="${var,,}"
+      val="${HOME}/.aws/${val//_/-}.env"
+      [[ -s "${val}" ]] || continue
+      declare -x "${var}=$(cat "${val}")"
+    done
+    # Verify that AWS credentials are functional.
+    aws sts get-caller-identity
+    # Log all PKCS11 helper activity, to simplify debugging.
+    export AWS_KMS_PKCS11_DEBUG=1
+    sb_key_source="aws"
+    shim_sign_key=(-c shim-sign-key -t shim-sign-key)
+    code_sign_key=(-c code-sign-key -t code-sign-key)
+  else
+    # Disable the PKCS11 helper.
+    rm /etc/pkcs11/modules/aws-kms-pkcs11.module
+
+    # Generate the PKCS12 archives for import.
+    openssl pkcs12 \
+      -export \
+      -passout pass: \
+      -inkey "${SBKEYS}/shim-sign.key" \
+      -in "${SBKEYS}/shim-sign.crt" \
+      -certfile "${SBKEYS}/db.crt" \
+      -out "${SBKEYS}/shim-sign.p12"
+
+    openssl pkcs12 \
+      -export \
+      -passout pass: \
+      -inkey "${SBKEYS}/code-sign.key" \
+      -in "${SBKEYS}/code-sign.crt" \
+      -certfile "${SBKEYS}/vendor.crt" \
+      -out "${SBKEYS}/code-sign.p12"
+
+    # Import certificates and private key archive.
+    pedb="/etc/pki/pesign"
+
+    certutil -d "${pedb}" -A -n db -i "${SBKEYS}/db.crt" -t ",,C"
+    certutil -d "${pedb}" -A -n shim-sign-key -i "${SBKEYS}/shim-sign.crt" -t ",,P"
+    pk12util -d "${pedb}" -i "${SBKEYS}/shim-sign.p12" -W ""
+
+    certutil -d "${pedb}" -A -n vendor -i "${SBKEYS}/vendor.crt" -t ",,C"
+    certutil -d "${pedb}" -A -n code-sign-key -i "${SBKEYS}/code-sign.crt" -t ",,P"
+    pk12util -d "${pedb}" -i "${SBKEYS}/code-sign.p12" -W ""
+
+    certutil -d "${pedb}" -L
+    sb_key_source="local"
+    shim_sign_key=(-c shim-sign-key)
+    code_sign_key=(-c code-sign-key)
+  fi
+
+  # Convert certificates from PEM format (ASCII) to DER (binary). This could be
+  # done when the certificates are created, but the resulting binary files are
+  # not as nice to store in source control.
+  for cert in PK KEK db vendor; do
+    openssl x509 \
+      -inform PEM -in "${SBKEYS}/${cert}.crt" \
+      -outform DER -out "${SBKEYS}/${cert}.cer"
+  done
+
+  # For signing the grub config, we need to embed the GPG public key in binary
+  # form, which is similarly awkward to store in source control.
+  gpg --batch --no-tty --import "${SBKEYS}/config-sign.key"
+  if [[ "${sb_key_source}" == "aws" ]]; then
+    gpg --batch --no-tty --card-status
+  fi
+  gpg --batch --no-tty --export >"${SBKEYS}/config-sign.pubkey"
+  gpg --batch --no-tty --list-keys
+}
+
+replace_vendor_certs() {
```

Comment:
Naming isn't great here - this isn't just doing "replace vendor certs", it's doing the entire signing operation for multiple artifacts.